swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll

```diff
@@ -277,35 +277,54 @@ abstract class RegExp extends Expr {
   private predicate isGroupStart(int i) { this.nonEscapedCharAt(i) = "(" and not this.inCharSet(i) }
 
   /**
-   * Holds if a parse mode starts between `start` and `end`.
+   * Holds if the initial part of a parse mode, not containing any
+   * mode characters is between `start` and `end`.
    */
-  private predicate flagGroupStart(int start, int end) {
+  private predicate flagGroupStartNoModes(int start, int end) {
     this.isGroupStart(start) and
     this.getChar(start + 1) = "?" and
     this.getChar(start + 2) in ["i", "x", "s", "m", "w"] and
     end = start + 2
   }
 
   /**
-   * Holds if a parse mode group is between `start` and `end`, and includes the
-   * mode flag `c`. For example the following span, with mode flag `i`:
+   * Holds if `pos` contains a mode character from the
+   * flag group starting at `start`.
+   */
+  private predicate modeCharacter(int start, int pos) {
+    this.flagGroupStartNoModes(start, pos)
+    or
+    this.modeCharacter(start, pos - 1) and
+    this.getChar(pos) in ["i", "x", "s", "m", "w"]
+  }
+
+  /**
+   * Holds if a parse mode group is between `start` and `end`.
+   */
+  private predicate flagGroupStart(int start, int end) {
+    this.flagGroupStartNoModes(start, _) and
+    end = max(int i | this.modeCharacter(start, i) | i + 1)
+  }
+
+  /**
+   * Holds if a parse mode group of this regex includes the mode flag `c`.
+   * For example the following parse mode group, with mode flag `i`:
```

Comment:
Nit: Should this be `"i"` instead of `i`, since it's a string literal rather than a symbol (in contrast with `c`)? (Not sure what the convention is for QL docs.)

Reply:
Well, `0` isn't a string, so I think that makes sense. I'd prefer both backticks (to show we're talking about code) and quotes (to show that the bit of code we're talking about is a string).

Basically, I think of it as an excerpt from a longer code fragment, either `flag(c)` (no quotes) or `flag("i")` (with quotes).

Reply:
Change made.